src/libraries/System.Runtime.InteropServices.JavaScript/gen/JsImportGenerator/JSExportGenerator.cs

```diff
@@ -0,0 +1,355 @@
+            var signatureContext = JSSignatureContext.Create(symbol, environment, generatorDiagnostics, ct);
+
+            var containingTypeContext = new ContainingSyntaxContext(originalSyntax);
+
+            var methodSyntaxTemplate = new ContainingSyntax(originalSyntax.Modifiers.StripTriviaFromTokens(), SyntaxKind.MethodDeclaration, originalSyntax.Identifier, originalSyntax.TypeParameterList);
+
+            return new IncrementalStubGenerationContext(
+                environment,
+                signatureContext,
+                containingTypeContext,
+                methodSyntaxTemplate,
+                new MethodSignatureDiagnosticLocations(originalSyntax),
+                jsExportData,
+                CreateGeneratorFactory(environment, options),
+                generatorDiagnostics.Diagnostics.ToImmutableArray());
+        }
+
+        private static MarshallingGeneratorFactoryKey<(TargetFramework, Version, JSGeneratorOptions)> CreateGeneratorFactory(StubEnvironment env, JSGeneratorOptions options)
+        {
+            IMarshallingGeneratorFactory generatorFactory;
+
+            generatorFactory = new UnsupportedMarshallingFactory();
+            JSGeneratorFactory jsGeneratorFactory = new JSGeneratorFactory(generatorFactory);
+
+            return MarshallingGeneratorFactoryKey.Create((env.TargetFramework, env.TargetFrameworkVersion, options), jsGeneratorFactory);
+        }
+
+        private static (MemberDeclarationSyntax, ImmutableArray<Diagnostic>) GenerateSource(
+            IncrementalStubGenerationContext incrementalContext,
+            JSGeneratorOptions options)
+        {
+            var diagnostics = new GeneratorDiagnostics();
+
+            try
+            {
+                // Generate stub code
+                var stubGenerator = new JSExportCodeGenerator(
+                incrementalContext.Environment,
+                incrementalContext.SignatureContext.ElementTypeInformation,
+                incrementalContext.JSExportData,
+                incrementalContext.SignatureContext,
+                (elementInfo, ex) =>
+                {
+                    diagnostics.ReportMarshallingNotSupported(incrementalContext.DiagnosticLocation, elementInfo, ex.NotSupportedDetails, ex.DiagnosticProperties ?? ImmutableDictionary<string, string>.Empty);
+                },
+                incrementalContext.GeneratorFactoryKey.GeneratorFactory);
+
+                BlockSyntax wrapper = stubGenerator.GenerateJSExportBody();
+                BlockSyntax registration = stubGenerator.GenerateJSExportRegistration();
+
+                return (PrintGeneratedSource(incrementalContext.StubMethodSyntaxTemplate, incrementalContext.SignatureContext, wrapper, registration), incrementalContext.Diagnostics.AddRange(diagnostics.Diagnostics));
+
+            }
+            catch (Exception ex)
+            {
+                var x = NamespaceDeclaration(IdentifierName("__failure__")).WithCloseBraceToken(Token(TriviaList(
```

Comment:
Just curious, is this correct way to handle errors in generators or is there anything better?